File: mRemoteV1/App/App.Update.vb
Imports System.IO
Imports System.Net
Imports System.ComponentModel
Imports System.Threading
Imports mRemoteNG.Tools
Imports System.Reflection

Namespace App
    Public Class Update
#Region "Events"
        Public Event GetUpdateInfoCompletedEvent As AsyncCompletedEventHandler
        Public Event GetChangeLogCompletedEvent As AsyncCompletedEventHandler
        Public Event GetAnnouncementInfoCompletedEvent As AsyncCompletedEventHandler
        Public Event DownloadUpdateProgressChangedEvent As DownloadProgressChangedEventHandler
        Public Event DownloadUpdateCompletedEvent As AsyncCompletedEventHandler
#End Region

#Region "Public Properties"
        Private _currentUpdateInfo As UpdateInfo
        Public ReadOnly Property CurrentUpdateInfo() As UpdateInfo
            Get
                Return _currentUpdateInfo
            End Get
        End Property

        Private _changeLog As String
        Public ReadOnly Property ChangeLog() As String
            Get
                Return _changeLog
            End Get
        End Property

        Private _currentAnnouncementInfo As AnnouncementInfo
        Public ReadOnly Property CurrentAnnouncementInfo() As AnnouncementInfo
            Get
                Return _currentAnnouncementInfo
            End Get
        End Property

        Public ReadOnly Property IsGetUpdateInfoRunning() As Boolean
            Get
                If _getUpdateInfoThread IsNot Nothing Then
                    If _getUpdateInfoThread.IsAlive Then Return True
                End If
                Return False
            End Get
        End Property

        Public ReadOnly Property IsGetChangeLogRunning() As Boolean
            Get
                If _getChangeLogThread IsNot Nothing Then
                    If _getChangeLogThread.IsAlive Then Return True
                End If
                Return False
            End Get
        End Property

        Public ReadOnly Property IsGetAnnouncementInfoRunning() As Boolean
            Get
                If _getAnnouncementInfoThread IsNot Nothing Then
                    If _getAnnouncementInfoThread.IsAlive Then Return True
                End If
                Return False
            End Get
        End Property

        Public ReadOnly Property IsDownloadUpdateRunning() As Boolean
            Get
                Return (_downloadUpdateWebClient IsNot Nothing)
            End Get
        End Property
#End Region

#Region "Public Methods"
        Public Sub New()
            SetProxySettings()
        End Sub

        Public Sub SetProxySettings()
            SetProxySettings(My.Settings.UpdateUseProxy, My.Settings.UpdateProxyAddress, My.Settings.UpdateProxyPort, My.Settings.UpdateProxyUseAuthentication, My.Settings.UpdateProxyAuthUser, Security.Crypt.Decrypt(My.Settings.UpdateProxyAuthPass, Info.General.EncryptionKey))
        End Sub

        Public Sub SetProxySettings(ByVal useProxy As Boolean, ByVal address As String, ByVal port As Integer, ByVal useAuthentication As Boolean, ByVal username As String, ByVal password As String)
            If useProxy And Not String.IsNullOrEmpty(address) Then
                If Not port = 0 Then
                    _webProxy = New WebProxy(address, port)
                Else
                    _webProxy = New WebProxy(address)
                End If

                If useAuthentication Then
                    _webProxy.Credentials = New NetworkCredential(username, password)
                Else
                    _webProxy.Credentials = Nothing
                End If
            Else
                _webProxy = Nothing
            End If
        End Sub

        Public Function IsUpdateAvailable() As Boolean
            If _currentUpdateInfo Is Nothing OrElse Not _currentUpdateInfo.IsValid Then Return False

            Return _currentUpdateInfo.Version > My.Application.Info.Version
        End Function

        Public Function IsAnnouncementAvailable() As Boolean
            If _currentAnnouncementInfo Is Nothing OrElse _
                (Not _currentAnnouncementInfo.IsValid Or _
                 String.IsNullOrEmpty(_currentAnnouncementInfo.Name)) Then Return False

            Return (Not _currentAnnouncementInfo.Name = My.Settings.LastAnnouncement)
        End Function

        Public Sub GetUpdateInfoAsync()
            If IsGetUpdateInfoRunning Then _getUpdateInfoThread.Abort()

            _getUpdateInfoThread = New Thread(AddressOf GetUpdateInfo)
            With _getUpdateInfoThread
                .SetApartmentState(ApartmentState.STA)
                .IsBackground = True
                .Start()
            End With
        End Sub

        Public Sub GetChangeLogAsync()
            If _currentUpdateInfo Is Nothing OrElse Not _currentUpdateInfo.IsValid Then
                Throw New InvalidOperationException("CurrentUpdateInfo is not valid. GetUpdateInfoAsync() must be called before calling GetChangeLogAsync().")
            End If

            If IsGetChangeLogRunning Then _getChangeLogThread.Abort()

            _getChangeLogThread = New Thread(AddressOf GetChangeLog)
            With _getChangeLogThread
                .SetApartmentState(ApartmentState.STA)
                .IsBackground = True
                .Start()
            End With
        End Sub

        Public Sub GetAnnouncementInfoAsync()
            If IsGetAnnouncementInfoRunning Then _getAnnouncementInfoThread.Abort()

            _getAnnouncementInfoThread = New Thread(AddressOf GetAnnouncementInfo)
            With _getAnnouncementInfoThread
                .SetApartmentState(ApartmentState.STA)
                .IsBackground = True
                .Start()
            End With
        End Sub

        Public Sub DownloadUpdateAsync()
            If _downloadUpdateWebClient IsNot Nothing Then
                Throw New InvalidOperationException("A previous call to DownloadUpdateAsync() is still in progress.")
            End If

            If _currentUpdateInfo Is Nothing OrElse Not _currentUpdateInfo.IsValid Then
                Throw New InvalidOperationException("CurrentUpdateInfo is not valid. GetUpdateInfoAsync() must be called before calling DownloadUpdateAsync().")
            End If

            _currentUpdateInfo.UpdateFilePath = Path.Combine(Path.GetTempPath(), Path.ChangeExtension(Path.GetRandomFileName, "exe"))
            DownloadUpdateWebClient.DownloadFileAsync(CurrentUpdateInfo.DownloadAddress, _currentUpdateInfo.UpdateFilePath)
        End Sub
#End Region

#Region "Private Properties"
        Private _downloadUpdateWebClient As WebClient
        Private ReadOnly Property DownloadUpdateWebClient() As WebClient
            Get
                If _downloadUpdateWebClient IsNot Nothing Then Return _downloadUpdateWebClient

                _downloadUpdateWebClient = CreateWebClient()

                AddHandler _downloadUpdateWebClient.DownloadProgressChanged, AddressOf DownloadUpdateProgressChanged
                AddHandler _downloadUpdateWebClient.DownloadFileCompleted, AddressOf DownloadUpdateCompleted

                Return _downloadUpdateWebClient
            End Get
        End Property
#End Region

#Region "Private Fields"
        Private _webProxy As WebProxy
        Private _getUpdateInfoThread As Thread
        Private _getChangeLogThread As Thread
        Private _getAnnouncementInfoThread As Thread
#End Region

#Region "Private Methods"
        Private Function CreateWebClient() As WebClient
            Dim webClient As New WebClient
            webClient.Headers.Add("user-agent", Info.General.UserAgent)
            webClient.Proxy = _webProxy
            Return webClient
        End Function

        Private Shared Function NewDownloadStringCompletedEventArgs(ByVal result As String, ByVal exception As Exception, ByVal cancelled As Boolean, ByVal userToken As Object) As DownloadStringCompletedEventArgs
            Dim type As Type = GetType(DownloadStringCompletedEventArgs)
            Const bindingFlags As BindingFlags = bindingFlags.NonPublic Or bindingFlags.Instance
            Dim argumentTypes() As Type = {GetType(String), GetType(Exception), GetType(Boolean), GetType(Object)}
            Dim constructor As ConstructorInfo = type.GetConstructor(bindingFlags, Nothing, argumentTypes, Nothing)
            Dim arguments() As Object = {result, exception, cancelled, userToken}

            Return constructor.Invoke(arguments)
        End Function

        Private Function DownloadString(ByVal address As Uri) As DownloadStringCompletedEventArgs
            Dim webClient As WebClient = CreateWebClient()
            Dim result As String = String.Empty
            Dim exception As Exception = Nothing
            Dim cancelled As Boolean = False

            Try
                result = webClient.DownloadString(address)
            Catch ex As ThreadAbortException
                cancelled = True
            Catch ex As Exception
                exception = ex
            End Try

            Return NewDownloadStringCompletedEventArgs(result, exception, cancelled, Nothing)
        End Function

        Private Sub GetUpdateInfo()
            Dim updateFileUri As New Uri(New Uri(My.Settings.UpdateAddress), New Uri(Info.Update.FileName, UriKind.Relative))
            Dim e As DownloadStringCompletedEventArgs = DownloadString(updateFileUri)

            If Not e.Cancelled And e.Error Is Nothing Then
                _currentUpdateInfo = UpdateInfo.FromString(e.Result)

                My.Settings.CheckForUpdatesLastCheck = Date.UtcNow
                If Not My.Settings.UpdatePending Then
                    My.Settings.UpdatePending = IsUpdateAvailable()
                End If
            End If

            RaiseEvent GetUpdateInfoCompletedEvent(Me, e)
        End Sub

        Private Sub GetChangeLog()
            Dim e As DownloadStringCompletedEventArgs = DownloadString(_currentUpdateInfo.ChangeLogAddress)

            If Not e.Cancelled And e.Error Is Nothing Then _changeLog = e.Result

            RaiseEvent GetChangeLogCompletedEvent(Me, e)
        End Sub

        Private Sub GetAnnouncementInfo()
            Dim announcementFileUri As New Uri(My.Settings.AnnouncementAddress)
            Dim e As DownloadStringCompletedEventArgs = DownloadString(announcementFileUri)

            If Not e.Cancelled And e.Error Is Nothing Then
                _currentAnnouncementInfo = AnnouncementInfo.FromString(e.Result)

                If Not String.IsNullOrEmpty(_currentAnnouncementInfo.Name) Then
                    My.Settings.LastAnnouncement = _currentAnnouncementInfo.Name
                End If
            End If

            RaiseEvent GetAnnouncementInfoCompletedEvent(Me, e)
        End Sub

        Private Sub DownloadUpdateProgressChanged(ByVal sender As Object, ByVal e As DownloadProgressChangedEventArgs)
            RaiseEvent DownloadUpdateProgressChangedEvent(sender, e)
        End Sub

        Private Sub DownloadUpdateCompleted(ByVal sender As Object, ByVal e As System.ComponentModel.AsyncCompletedEventArgs)
            Dim raiseEventArgs As AsyncCompletedEventArgs = e

            If Not e.Cancelled And e.Error Is Nothing Then
                Try
                    Dim updateAuthenticode As New Authenticode(_currentUpdateInfo.UpdateFilePath)
                    With updateAuthenticode
                        .RequireThumbprintMatch = True
                        .ThumbprintToMatch = _currentUpdateInfo.CertificateThumbprint

                        If Not .Verify() = Authenticode.StatusValue.Verified Then
                            If .Status = Authenticode.StatusValue.UnhandledException Then
                                Throw .Exception
                            Else
                                Throw New Exception(.StatusMessage)
                            End If
                        End If
                    End With
                Catch ex As Exception
                    raiseEventArgs = New AsyncCompletedEventArgs(ex, False, Nothing)
                End Try
            End If

            If raiseEventArgs.Cancelled Or raiseEventArgs.Error IsNot Nothing Then
                File.Delete(_currentUpdateInfo.UpdateFilePath)
            End If

            RaiseEvent DownloadUpdateCompletedEvent(Me, raiseEventArgs)

            _downloadUpdateWebClient.Dispose()
            _downloadUpdateWebClient = Nothing
        End Sub
#End Region

#Region "Public Classes"
        Public Class UpdateInfo
#Region "Public Properties"
            Public Property IsValid As Boolean
            Public Property Version As Version
            Public Property DownloadAddress As Uri
            Public Property UpdateFilePath As String
            Public Property ChangeLogAddress As Uri
            Public Property ImageAddress As Uri
            Public Property ImageLinkAddress As Uri
            Public Property CertificateThumbprint As String
#End Region

#Region "Public Methods"
            Public Shared Function FromString(ByVal input As String) As UpdateInfo
                Dim newInfo As New UpdateInfo
                With newInfo
                    If String.IsNullOrEmpty(input) Then
                        .IsValid = False
                    Else
                        Dim updateFile As New UpdateFile(input)
                        .Version = updateFile.GetVersion("Version")
                        .DownloadAddress = updateFile.GetUri("dURL")
                        .ChangeLogAddress = updateFile.GetUri("clURL")
                        .ImageAddress = updateFile.GetUri("imgURL")
                        .ImageLinkAddress = updateFile.GetUri("imgURLLink")
                        .CertificateThumbprint = updateFile.GetThumbprint("CertificateThumbprint")
                        .IsValid = True
                    End If
                End With
                Return newInfo
            End Function
#End Region
        End Class

        Public Class AnnouncementInfo
#Region "Public Properties"
            Public Property IsValid As Boolean
            Public Property Name As String
            Public Property Address As Uri
#End Region

#Region "Public Methods"
            Public Shared Function FromString(ByVal input As String) As AnnouncementInfo
                Dim newInfo As New AnnouncementInfo
                With newInfo
                    If String.IsNullOrEmpty(input) Then
                        .IsValid = False
                    Else
                        Dim updateFile As New UpdateFile(input)
                        .Name = updateFile.GetString("Name")
                        .Address = updateFile.GetUri("URL")
                        .IsValid = True
                    End If
                End With
                Return newInfo
            End Function
#End Region
        End Class
#End Region

#Region "Private Classes"
        Private Class UpdateFile
#Region "Public Properties"
            Private ReadOnly _items As New Dictionary(Of String, String)(StringComparer.InvariantCultureIgnoreCase)
            ' ReSharper disable MemberCanBePrivate.Local
            Public ReadOnly Property Items() As Dictionary(Of String, String)
                ' ReSharper restore MemberCanBePrivate.Local
                Get
                    Return _items
                End Get
            End Property
#End Region

#Region "Public Methods"
            Public Sub New(ByVal content As String)
                FromString(content)
            End Sub

            ' ReSharper disable MemberCanBePrivate.Local
            Public Sub FromString(ByVal content As String)
                ' ReSharper restore MemberCanBePrivate.Local
                If String.IsNullOrEmpty(content) Then
                Else
                    Dim lineSeparators() As Char = New Char() {Chr(&HA), Chr(&HD)}
                    Dim keyValueSeparators() As Char = New Char() {":", "="}
                    Dim commentCharacters() As Char = New Char() {"#", ";", "'"}

                    Dim lines() As String = content.Split(lineSeparators, StringSplitOptions.RemoveEmptyEntries)
                    For Each line As String In lines
                        line = line.Trim()
                        If line.Length = 0 Then Continue For
                        If Not line.Substring(0, 1).IndexOfAny(commentCharacters) = -1 Then Continue For

                        Dim parts() As String = line.Split(keyValueSeparators, 2)
                        If Not parts.Length = 2 Then Continue For
                        Dim key As String = parts(0).Trim()
                        Dim value As String = parts(1).Trim()

                        _items.Add(key, value)
                    Next
                End If
            End Sub

            ' ReSharper disable MemberCanBePrivate.Local
            Public Function GetString(ByVal key As String) As String
                ' ReSharper restore MemberCanBePrivate.Local
                If Not Items.ContainsKey(key) Then Return String.Empty
                Return Items(key)
            End Function

            Public Function GetVersion(ByVal key As String) As Version
                Dim value As String = GetString(key)
                If String.IsNullOrEmpty(value) Then Return Nothing
                Return New Version(value)
            End Function

            Public Function GetUri(ByVal key As String) As Uri
                Dim value As String = GetString(key)
                If String.IsNullOrEmpty(value) Then Return Nothing
                Return New Uri(value)
            End Function

            Public Function GetThumbprint(ByVal key As String) As String
                Return GetString(key).Replace(" ", "").ToUpperInvariant()
            End Function
#End Region
        End Class
#End Region
    End Class
End Namespace

File: mRemoteV1/App/App.Update.cs
using System.Collections.Generic;
using System;
using System.IO;
using System.Net;
using System.ComponentModel;
using System.Threading;
using mRemoteNG.Tools;
using System.Reflection;


namespace mRemoteNG.App
{
	public class Update
	{
        #region Events
		private AsyncCompletedEventHandler GetUpdateInfoCompletedEventEvent;
		public event AsyncCompletedEventHandler GetUpdateInfoCompletedEvent
		{
			add
			{
				GetUpdateInfoCompletedEventEvent = (AsyncCompletedEventHandler) System.Delegate.Combine(GetUpdateInfoCompletedEventEvent, value);
			}
			remove
			{
				GetUpdateInfoCompletedEventEvent = (AsyncCompletedEventHandler) System.Delegate.Remove(GetUpdateInfoCompletedEventEvent, value);
			}
		}
			
		private AsyncCompletedEventHandler GetChangeLogCompletedEventEvent;
		public event AsyncCompletedEventHandler GetChangeLogCompletedEvent
		{
			add
			{
				GetChangeLogCompletedEventEvent = (AsyncCompletedEventHandler) System.Delegate.Combine(GetChangeLogCompletedEventEvent, value);
			}
			remove
			{
				GetChangeLogCompletedEventEvent = (AsyncCompletedEventHandler) System.Delegate.Remove(GetChangeLogCompletedEventEvent, value);
			}
		}
			
		private AsyncCompletedEventHandler GetAnnouncementInfoCompletedEventEvent;
		public event AsyncCompletedEventHandler GetAnnouncementInfoCompletedEvent
		{
			add
			{
				GetAnnouncementInfoCompletedEventEvent = (AsyncCompletedEventHandler) System.Delegate.Combine(GetAnnouncementInfoCompletedEventEvent, value);
			}
			remove
			{
				GetAnnouncementInfoCompletedEventEvent = (AsyncCompletedEventHandler) System.Delegate.Remove(GetAnnouncementInfoCompletedEventEvent, value);
			}
		}
			
		private DownloadProgressChangedEventHandler DownloadUpdateProgressChangedEventEvent;
		public event DownloadProgressChangedEventHandler DownloadUpdateProgressChangedEvent
		{
			add
			{
				DownloadUpdateProgressChangedEventEvent = (DownloadProgressChangedEventHandler) System.Delegate.Combine(DownloadUpdateProgressChangedEventEvent, value);
			}
			remove
			{
				DownloadUpdateProgressChangedEventEvent = (DownloadProgressChangedEventHandler) System.Delegate.Remove(DownloadUpdateProgressChangedEventEvent, value);
			}
		}
			
		private AsyncCompletedEventHandler DownloadUpdateCompletedEventEvent;
		public event AsyncCompletedEventHandler DownloadUpdateCompletedEvent
		{
			add
			{
				DownloadUpdateCompletedEventEvent = (AsyncCompletedEventHandler) System.Delegate.Combine(DownloadUpdateCompletedEventEvent, value);
			}
			remove
			{
				DownloadUpdateCompletedEventEvent = (AsyncCompletedEventHandler) System.Delegate.Remove(DownloadUpdateCompletedEventEvent, value);
			}
		}
        #endregion
			
        #region Public Properties
		private UpdateInfo _currentUpdateInfo;
        public UpdateInfo CurrentUpdateInfo
		{
			get
			{
				return _currentUpdateInfo;
			}
		}
			
		private string _changeLog;
        public string ChangeLog
		{
			get
			{
				return _changeLog;
			}
		}
			
		private AnnouncementInfo _currentAnnouncementInfo;
        public AnnouncementInfo CurrentAnnouncementInfo
		{
			get
			{
				return _currentAnnouncementInfo;
			}
		}
			
        public bool IsGetUpdateInfoRunning
		{
			get
			{
				if (_getUpdateInfoThread != null)
				{
					if (_getUpdateInfoThread.IsAlive)
					{
						return true;
					}
				}
				return false;
			}
		}
			
        public bool IsGetChangeLogRunning
		{
			get
			{
				if (_getChangeLogThread != null)
				{
					if (_getChangeLogThread.IsAlive)
					{
						return true;
					}
				}
				return false;
			}
		}
			
        public bool IsGetAnnouncementInfoRunning
		{
			get
			{
				if (_getAnnouncementInfoThread != null)
				{
					if (_getAnnouncementInfoThread.IsAlive)
					{
						return true;
					}
				}
				return false;
			}
		}
			
        public bool IsDownloadUpdateRunning
		{
			get
			{
				return (_downloadUpdateWebClient != null);
			}
		}
        #endregion
			
        #region Public Methods
		public Update()
		{
			SetProxySettings();
		}
			
		public void SetProxySettings()
		{
			SetProxySettings(System.Convert.ToBoolean(My.Settings.Default.UpdateUseProxy), System.Convert.ToString(My.Settings.Default.UpdateProxyAddress), System.Convert.ToInt32(My.Settings.Default.UpdateProxyPort), System.Convert.ToBoolean(My.Settings.Default.UpdateProxyUseAuthentication), System.Convert.ToString(My.Settings.Default.UpdateProxyAuthUser), Security.Crypt.Decrypt(System.Convert.ToString(My.Settings.Default.UpdateProxyAuthPass), Info.General.EncryptionKey));
		}
			
		public void SetProxySettings(bool useProxy, string address, int port, bool useAuthentication, string username, string password)
		{
			if (useProxy && !string.IsNullOrEmpty(address))
			{
				if (!(port == 0))
				{
					_webProxy = new WebProxy(address, port);
				}
				else
				{
					_webProxy = new WebProxy(address);
				}
					
				if (useAuthentication)
				{
					_webProxy.Credentials = new NetworkCredential(username, password);
				}
				else
				{
					_webProxy.Credentials = null;
				}
			}
			else
			{
				_webProxy = null;
			}
		}
			
		public bool IsUpdateAvailable()
		{
			if (_currentUpdateInfo == null || !_currentUpdateInfo.IsValid)
			{
				return false;
			}
				
			return _currentUpdateInfo.Version > (new Microsoft.VisualBasic.ApplicationServices.WindowsFormsApplicationBase()).Info.Version;
		}
			
		public bool IsAnnouncementAvailable()
		{
			if (_currentAnnouncementInfo == null || (!_currentAnnouncementInfo.IsValid || string.IsNullOrEmpty(_currentAnnouncementInfo.Name)))
			{
				return false;
			}
				
			return !(_currentAnnouncementInfo.Name == My.Settings.Default.LastAnnouncement);
		}
			
		public void GetUpdateInfoAsync()
		{
			if (IsGetUpdateInfoRunning)
			{
				_getUpdateInfoThread.Abort();
			}
				
			_getUpdateInfoThread = new Thread(new System.Threading.ThreadStart(GetUpdateInfo));
			_getUpdateInfoThread.SetApartmentState(ApartmentState.STA);
			_getUpdateInfoThread.IsBackground = true;
			_getUpdateInfoThread.Start();
		}
			
		public void GetChangeLogAsync()
		{
			if (_currentUpdateInfo == null || !_currentUpdateInfo.IsValid)
			{
				throw (new InvalidOperationException("CurrentUpdateInfo is not valid. GetUpdateInfoAsync() must be called before calling GetChangeLogAsync()."));
			}
				
			if (IsGetChangeLogRunning)
			{
				_getChangeLogThread.Abort();
			}
				
			_getChangeLogThread = new Thread(new System.Threading.ThreadStart(GetChangeLog));
			_getChangeLogThread.SetApartmentState(ApartmentState.STA);
			_getChangeLogThread.IsBackground = true;
			_getChangeLogThread.Start();
		}
			
		public void GetAnnouncementInfoAsync()
		{
			if (IsGetAnnouncementInfoRunning)
			{
				_getAnnouncementInfoThread.Abort();
			}
				
			_getAnnouncementInfoThread = new Thread(new System.Threading.ThreadStart(GetAnnouncementInfo));
			_getAnnouncementInfoThread.SetApartmentState(ApartmentState.STA);
			_getAnnouncementInfoThread.IsBackground = true;
			_getAnnouncementInfoThread.Start();
		}
			
		public void DownloadUpdateAsync()
		{
			if (_downloadUpdateWebClient != null)
			{
				throw (new InvalidOperationException("A previous call to DownloadUpdateAsync() is still in progress."));
			}
				
			if (_currentUpdateInfo == null || !_currentUpdateInfo.IsValid)
			{
				throw (new InvalidOperationException("CurrentUpdateInfo is not valid. GetUpdateInfoAsync() must be called before calling DownloadUpdateAsync()."));
			}
				
			_currentUpdateInfo.UpdateFilePath = Path.Combine(Path.GetTempPath(), Path.ChangeExtension(Path.GetRandomFileName(), "exe"));
			DownloadUpdateWebClient.DownloadFileAsync(CurrentUpdateInfo.DownloadAddress, _currentUpdateInfo.UpdateFilePath);
		}
        #endregion
			
        #region Private Properties
		private WebClient _downloadUpdateWebClient;
        private WebClient DownloadUpdateWebClient
		{
			get
			{
				if (_downloadUpdateWebClient != null)
				{
					return _downloadUpdateWebClient;
				}
					
				_downloadUpdateWebClient = CreateWebClient();
					
				_downloadUpdateWebClient.DownloadProgressChanged += DownloadUpdateProgressChanged;
				_downloadUpdateWebClient.DownloadFileCompleted += DownloadUpdateCompleted;
					
				return _downloadUpdateWebClient;
			}
		}
        #endregion
			
        #region Private Fields
		private WebProxy _webProxy;
		private Thread _getUpdateInfoThread;
		private Thread _getChangeLogThread;
		private Thread _getAnnouncementInfoThread;
        #endregion
			
        #region Private Methods
		private WebClient CreateWebClient()
		{
			WebClient webClient = new WebClient();
			webClient.Headers.Add("user-agent", Info.General.UserAgent);
			webClient.Proxy = _webProxy;
			return webClient;
		}
			
		private static DownloadStringCompletedEventArgs NewDownloadStringCompletedEventArgs(string result, Exception exception, bool cancelled, object userToken)
		{
			Type type = typeof(DownloadStringCompletedEventArgs);
            const BindingFlags bindingFlags = BindingFlags.NonPublic | BindingFlags.Instance;
			Type[] argumentTypes = new Type[] {typeof(string), typeof(Exception), typeof(bool), typeof(object)};
			ConstructorInfo constructor = type.GetConstructor(bindingFlags, null, argumentTypes, null);
			object[] arguments = new object[] {result, exception, cancelled, userToken};

            return (DownloadStringCompletedEventArgs)constructor.Invoke(arguments);
		}
			
		private DownloadStringCompletedEventArgs DownloadString(Uri address)
		{
			WebClient webClient = CreateWebClient();
			string result = string.Empty;
			Exception exception = null;
			bool cancelled = false;
				
			try
			{
				result = webClient.DownloadString(address);
			}
			catch (ThreadAbortException)
			{
				cancelled = true;
			}
			catch (Exception ex)
			{
				exception = ex;
			}
				
			return NewDownloadStringCompletedEventArgs(result, exception, cancelled, null);
		}
			
		private void GetUpdateInfo()
		{
			Uri updateFileUri = new Uri(new Uri(System.Convert.ToString(My.Settings.Default.UpdateAddress)), new Uri(Info.Update.FileName, UriKind.Relative));
			DownloadStringCompletedEventArgs e = DownloadString(updateFileUri);
				
			if (!e.Cancelled && e.Error == null)
			{
				_currentUpdateInfo = UpdateInfo.FromString(e.Result);
					
				My.Settings.Default.CheckForUpdatesLastCheck = DateTime.UtcNow;
				if (!My.Settings.Default.UpdatePending)
				{
					My.Settings.Default.UpdatePending = IsUpdateAvailable();
				}
			}
				
			if (GetUpdateInfoCompletedEventEvent != null)
				GetUpdateInfoCompletedEventEvent(this, e);
		}
			
		private void GetChangeLog()
		{
			DownloadStringCompletedEventArgs e = DownloadString(_currentUpdateInfo.ChangeLogAddress);
				
			if (!e.Cancelled && e.Error == null)
			{
				_changeLog = e.Result;
			}
				
			if (GetChangeLogCompletedEventEvent != null)
				GetChangeLogCompletedEventEvent(this, e);
		}
			
		private void GetAnnouncementInfo()
		{
			Uri announcementFileUri = new Uri(System.Convert.ToString(My.Settings.Default.AnnouncementAddress));
			DownloadStringCompletedEventArgs e = DownloadString(announcementFileUri);
				
			if (!e.Cancelled && e.Error == null)
			{
				_currentAnnouncementInfo = AnnouncementInfo.FromString(e.Result);
					
				if (!string.IsNullOrEmpty(_currentAnnouncementInfo.Name))
				{
					My.Settings.Default.LastAnnouncement = _currentAnnouncementInfo.Name;
				}
			}
				
			if (GetAnnouncementInfoCompletedEventEvent != null)
				GetAnnouncementInfoCompletedEventEvent(this, e);
		}
			
		private void DownloadUpdateProgressChanged(object sender, DownloadProgressChangedEventArgs e)
		{
			if (DownloadUpdateProgressChangedEventEvent != null)
				DownloadUpdateProgressChangedEventEvent(sender, e);
		}
			
		private void DownloadUpdateCompleted(object sender, System.ComponentModel.AsyncCompletedEventArgs e)
		{
			AsyncCompletedEventArgs raiseEventArgs = e;
				
			if (!e.Cancelled && e.Error == null)
			{
				try
				{
					Authenticode updateAuthenticode = new Authenticode(_currentUpdateInfo.UpdateFilePath);
					updateAuthenticode.RequireThumbprintMatch = true;
					updateAuthenticode.ThumbprintToMatch = _currentUpdateInfo.CertificateThumbprint;
						
					if (!(updateAuthenticode.Verify() == Authenticode.StatusValue.Verified))
					{
						if (updateAuthenticode.Status == Authenticode.StatusValue.UnhandledException)
						{
							throw (updateAuthenticode.Exception);
						}
						else
						{
							throw (new Exception(updateAuthenticode.StatusMessage));
						}
					}
				}
				catch (Exception ex)
				{
					raiseEventArgs = new AsyncCompletedEventArgs(ex, false, null);
				}
			}
				
			if (raiseEventArgs.Cancelled || raiseEventArgs.Error != null)
			{
				File.Delete(_currentUpdateInfo.UpdateFilePath);
			}
				
			if (DownloadUpdateCompletedEventEvent != null)
				DownloadUpdateCompletedEventEvent(this, raiseEventArgs);
				
			_downloadUpdateWebClient.Dispose();
			_downloadUpdateWebClient = null;
		}
        #endregion
			
        #region Public Classes
		public class UpdateInfo
		{
            #region Public Properties
			public bool IsValid {get; set;}
			public Version Version {get; set;}
			public Uri DownloadAddress {get; set;}
			public string UpdateFilePath {get; set;}
			public Uri ChangeLogAddress {get; set;}
			public Uri ImageAddress {get; set;}
			public Uri ImageLinkAddress {get; set;}
			public string CertificateThumbprint {get; set;}
            #endregion
				
            #region Public Methods
			public static UpdateInfo FromString(string input)
			{
				UpdateInfo newInfo = new UpdateInfo();
				if (string.IsNullOrEmpty(input))
				{
					newInfo.IsValid = false;
				}
				else
				{
					UpdateFile updateFile = new UpdateFile(input);
					newInfo.Version = updateFile.GetVersion("Version");
					newInfo.DownloadAddress = updateFile.GetUri("dURL");
					newInfo.ChangeLogAddress = updateFile.GetUri("clURL");
					newInfo.ImageAddress = updateFile.GetUri("imgURL");
					newInfo.ImageLinkAddress = updateFile.GetUri("imgURLLink");
					newInfo.CertificateThumbprint = updateFile.GetThumbprint("CertificateThumbprint");
					newInfo.IsValid = true;
				}
				return newInfo;
			}
            #endregion
		}
			
		public class AnnouncementInfo
		{
            #region Public Properties
			public bool IsValid {get; set;}
			public string Name {get; set;}
			public Uri Address {get; set;}
            #endregion
				
            #region Public Methods
			public static AnnouncementInfo FromString(string input)
			{
				AnnouncementInfo newInfo = new AnnouncementInfo();
				if (string.IsNullOrEmpty(input))
				{
					newInfo.IsValid = false;
				}
				else
				{
					UpdateFile updateFile = new UpdateFile(input);
					newInfo.Name = updateFile.GetString("Name");
					newInfo.Address = updateFile.GetUri("URL");
					newInfo.IsValid = true;
				}
				return newInfo;
			}
            #endregion
		}
        #endregion
			
        #region Private Classes
		private class UpdateFile
		{
            #region Public Properties
			private Dictionary<string, string> _items = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
			// ReSharper disable MemberCanBePrivate.Local
            public Dictionary<string, string> Items
			{
				// ReSharper restore MemberCanBePrivate.Local
				get
				{
					return _items;
				}
			}
            #endregion
				
            #region Public Methods
			public UpdateFile(string content)
			{
				FromString(content);
			}
				
			// ReSharper disable MemberCanBePrivate.Local
			public void FromString(string content)
			{
				// ReSharper restore MemberCanBePrivate.Local
				if (string.IsNullOrEmpty(content))
				{
				}
				else
				{
					char[] lineSeparators = new char[] {'\n', '\r'};
					char[] keyValueSeparators = new char[] {':', '='};
					char[] commentCharacters = new char[] {'#', ';', '\''};
						
					string[] lines = content.Split(lineSeparators.ToString().ToCharArray(), StringSplitOptions.RemoveEmptyEntries);
					foreach (string line in lines)
					{
						string trimmedLine = line.Trim();
                        if (trimmedLine.Length == 0)
						{
							continue;
						}
                        if (!(trimmedLine.Substring(0, 1).IndexOfAny(commentCharacters.ToString().ToCharArray()) == -1))
						{
							continue;
						}

                        string[] parts = trimmedLine.Split(keyValueSeparators.ToString().ToCharArray(), 2);
						if (!(parts.Length == 2))
						{
							continue;
						}
						string key = System.Convert.ToString(parts[0].Trim());
						string value = System.Convert.ToString(parts[1].Trim());
							
						_items.Add(key, value);
					}
				}
			}
				
			// ReSharper disable MemberCanBePrivate.Local
			public string GetString(string key)
			{
				// ReSharper restore MemberCanBePrivate.Local
				if (!Items.ContainsKey(key))
				{
					return string.Empty;
				}
				return this._items[key];
			}
				
			public Version GetVersion(string key)
			{
				string value = GetString(key);
				if (string.IsNullOrEmpty(value))
				{
					return null;
				}
				return new Version(value);
			}
				
			public Uri GetUri(string key)
			{
				string value = GetString(key);
				if (string.IsNullOrEmpty(value))
				{
					return null;
				}
				return new Uri(value);
			}
				
			public string GetThumbprint(string key)
			{
				return GetString(key).Replace(" ", "").ToUpperInvariant();
			}
        #endregion
		}
        #endregion
	}
}
